src/Lecture2.hs

```diff
@@ -1,170 +1,268 @@
-    ) where
+  ( -- * Normal
+    lazyProduct,
+    duplicate,
+    removeAt,
+    evenLists,
+    dropSpaces,
+    Knight (..),
+    dragonFight,
+
+    -- * Hard
+    isIncreasing,
+    merge,
+    mergeSort,
+    Expr (..),
+    Variables,
+    EvalError (..),
+    eval,
+    constantFolding,
+  )
+where
 
 -- VVV If you need to import libraries, do it after this line ... VVV
 
--- ^^^ and before this line. Otherwise the test suite might fail  ^^^
+import Data.Char (isSpace)
 
-{- | Implement a function that finds a product of all the numbers in
-the list. But implement a lazier version of this function: if you see
-zero, you can stop calculating product and return 0 immediately.
+-- ^ ^^ and before this line. Otherwise the test suite might fail  ^^^
 
->>> lazyProduct [4, 3, 7]
-84
--}
+-- | Implement a function that finds a product of all the numbers in
+-- the list. But implement a lazier version of this function: if you see
+-- zero, you can stop calculating product and return 0 immediately.
+--
+-- >>> lazyProduct [4, 3, 7]
+-- 84
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
-
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
+lazyProduct lst = case lst of
+  [] -> 1
+  (0 : _) -> 0
+  [x] -> x
+  (x : xs) -> x * lazyProduct xs
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
-
-{- | Implement function that takes index and a list and removes the
-element at the given position. Additionally, this function should also
-return the removed element.
-
->>> removeAt 0 [1 .. 5]
-(Just 1,[2,3,4,5])
-
->>> removeAt 10 [1 .. 5]
-(Nothing,[1,2,3,4,5])
--}
-removeAt = error "TODO"
-
-{- | Write a function that takes a list of lists and returns only
-lists of even lengths.
-
->>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
-[[3,1,2,7],[]]
-
-♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
-  in this function.
--}
-evenLists = error "TODO"
+duplicate lst = case lst of
+  [] -> []
+  (x : xs) -> x : x : duplicate xs
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt = removeHelper []
+  where
+    -- cannot use length, lst can be infinite
+    removeHelper :: [a] -> Int -> [a] -> (Maybe a, [a])
+    removeHelper before index after
+      | index < 0 = (Nothing, after)
+      | null after = (Nothing, before)
+      | length before == index = (Just (head after), before ++ drop 1 after)
+      | otherwise = removeHelper (before ++ take 1 after) index (drop 1 after)
```

Comment:
This is a valid implementation 👍🏻 

It has a few inefficient places:

* You compare `index` with 0 on every recursive step where it's enough to compare only once
* Calculating `length` of the `before` list on every recursive step (and `length` is an expensive operation for lists)
* Adding an element to the end of `before` list

You can implement this function without these two drawbacks in the following more efficient way:

```suggestion
removeAt :: Int -> [a] -> (Maybe a, [a])
removeAt n l
    | n < 0 = (Nothing, l)
    | otherwise = go n l
  where
    go _ [] = (Nothing, [])
    go 0 (x : xs) = (Just x, xs)
    go i (x : xs) =
        let (val, rest) = go (i - 1) xs
        in (val, x : rest)
```